draft-jmiller-json-proof-token.md

```diff
@@ -0,0 +1,148 @@
+
+The most significant divergence required by JPTs is that of supporting values that may be disclosed or may instead only be a proof about the value.  Applications are required to interact with the JPT on a claim-by-claim basis instead of just verifying a JWT and then being able to interact with the JSON body directly.
+
+## Proofs
+
+In order to generate a variety of ZKPs of knowledge, range, membeship, or other predicates it is essential that the individual payloads are singular values.  This greatly simplifies the task of linking a derived proof of a given claim to the specific payload that was also signed by the issuer.
+
+While JPTs definitely support protected claims that have more complex object or array compound values, it must also allow and encourage protecting simple claim values individually in a single payload such as strings, numbers, and booleans.
+
+# Layout
+
+A JSON Proof Token is defined as the combination of two different parts: the layout and the payloads. Given the design considerations the simplest solution is to move the token's encompassing JSON body to an external shared definition while preserving only the individual claim values in the JWP container.
+
+The layout definition MUST be accessible to and known by all parties that will process the JPT.  Since this is also true of the public key, the ideal location for the layout definition is in the JWK itself. (Note: layout definition could move to a standalone JSON Proof Key specification)
+
+A layout is the symbolic equivalent to the JSON body of a JWT, containing all of the claim names along with any additional structures or fixed metadata.  The individual claim values are not included in the layout and instead are replaced by payload references.
+
+These claim value references can be present in two different types: direct and indirect.  A direct reference is always an array with a single integer value, acting as an index into the array of payloads contained in the JPT.  All payloads referenced directly MUST be a string representing a JSON value.
+
+Indirect references are those understood only in the context of supporting applications.  Payloads with indirect references will have content types that vary by the given context and may be textual or binary.  This allows JPTs to include raw values without incurring the penalty of a double-encoding into a JSON-safe string.
+
+Layout definition object, using single-integer-value arrays as the index references to which payload:
+```json
+{
+  "kty": "ZKP",
+  "crv": "Bls12381G2",
+  "x": "qM4Gi4razIIAXpDSlHB7-pPoo6GOChoBbSLxr7rNwb8mxyVykbmKQGNb0kI7iegDAs9cIwf6DAsCGi7BVs48MG-iw4PsP0L136g2gQpZjrKsr4GbkV5EIx0R2BjIJNfQ",
+  "kid": "HjfcpyjuZQ-O8Ye2hQnNbT9RbbnrobptdnExR0DUjU8",
+  "alg": "BBS",
+  "use": "proof",
+  "lyt": {
```

Comment:
this is layout right? are other layouts configurable, or is this the only supported one?

Reply:
Interested to see this debate over "key" vs "header" embedding for this value.

I tend to have a natural frustation with binding keys to specifc uses arbitrarily.

Mostly because it violates the principle of "if it can happen, it does happen"...

You can say this key is "signing" and not "encryption" but if its supports ECHD, you can't stop folks from doing that....

We see this play out over and over again... most recently in the did core concept of "verification relationships"... or "keys used for specific purposes like authentication or credential issuance".

This feels even worse than that to me, since its "keys used to issue specific credential types".... I'd rather not see that zkp-cl feature repeated here.

Instead I would like to see a system where the same key, might be for many purposes, without revealing what purposes it was used for... for example:


```
{
      "kty": "EC",
      "crv": "P-256",
      "x": "KSbFyuU2BC1erMom2dMnQM5X4ifBdDXm1yQujBlGbrM",
      "y": "e50tXUMdJkdPseZkuIgCR_FMZ_zozeKusq6W4tr6-2M"
}
```

I use it for encryption, I use it for signing, I use it for merkle proofs, you don't know any of that by seeing the key.

Reply:
Another consideration is that you might want to encode `lyt` by reference instead of by value....

Consider a commitment to salted hash (lyt)... etc... it might be that there are really large potential space savings for large "payload types"... when you don't need to keep repeating the payload structure in the header or the key... btw this is what linked data contexts are supposed to help you do... maybe they didn't solve the problem the best way... but imo its still worth considering the concept of an `lyt` registry... since someone will create that, regardless of what we want.